File: wnode-sdk-go/integrations/registry.go
package integrations

import (
	"errors"
	"sort"
)

type IntegrationMetadata struct {
	Name               string             `json:"name"`
	Version            string             `json:"version"`
	Capabilities       CapabilitySet      `json:"capabilities"`
	DeterminismProfile DeterminismProfile `json:"determinismProfile"`
	SecurityProfile    SecurityProfile    `json:"securityProfile"`
}

type IntegrationRegistry struct {
	adapters map[string]IntegrationAdapter
}

func NewIntegrationRegistry() *IntegrationRegistry {
	return &IntegrationRegistry{
		adapters: make(map[string]IntegrationAdapter),
	}
}

func (r *IntegrationRegistry) RegisterIntegration(adapter IntegrationAdapter) error {
	name := adapter.Name()
	if _, exists := r.adapters[name]; exists {
		return errors.New("integration already registered")
	}

	profile := adapter.DeterminismProfile()
	if !profile.IsPurelyDeterministic && profile.ReliesOnRandomness {
		return errors.New("integration violates strict determinism invariants")
	}

	r.adapters[name] = adapter
	return nil
}

func (r *IntegrationRegistry) GetIntegration(name string) (IntegrationAdapter, error) {
	adapter, exists := r.adapters[name]
	if !exists {
		return nil, errors.New("integration not found")
	}
	return adapter, nil
}

func (r *IntegrationRegistry) ListIntegrations() []IntegrationMetadata {
	keys := make([]string, 0, len(r.adapters))
	for k := range r.adapters {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	list := make([]IntegrationMetadata, len(keys))
	for i, k := range keys {
		adapter := r.adapters[k]
		list[i] = IntegrationMetadata{
			Name:               adapter.Name(),
			Version:            adapter.Version(),
			Capabilities:       adapter.Capabilities(),
			DeterminismProfile: adapter.DeterminismProfile(),
			SecurityProfile:    adapter.SecurityProfile(),
		}
	}
	return list
}

File: packages/wnode-sdk-ts/src/integrations/registry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IntegrationRegistry = void 0;
class IntegrationRegistry {
    adapters = new Map();
    registerIntegration(adapter) {
        if (this.adapters.has(adapter.name)) {
            throw new Error(`Integration ${adapter.name} is already registered.`);
        }
        // Basic validation of determinism profile
        const profile = adapter.determinismProfile();
        if (!profile.isPurelyDeterministic && profile.reliesOnRandomness) {
            throw new Error(`Integration ${adapter.name} violates strict determinism invariants (relies on randomness).`);
        }
        this.adapters.set(adapter.name, adapter);
    }
    getIntegration(name) {
        const adapter = this.adapters.get(name);
        if (!adapter) {
            throw new Error(`Integration ${name} not found.`);
        }
        return adapter;
    }
    listIntegrations() {
        // Return sorted deterministically by name
        const keys = Array.from(this.adapters.keys()).sort();
        return keys.map(key => {
            const adapter = this.adapters.get(key);
            return {
                name: adapter.name,
                version: adapter.version,
                capabilities: adapter.capabilities(),
                determinismProfile: adapter.determinismProfile(),
                securityProfile: adapter.securityProfile()
            };
        });
    }
    validateIntegrationCapabilities(name, requiredCaps) {
        const adapter = this.getIntegration(name);
        const caps = adapter.capabilities();
        if (requiredCaps.canFetch !== undefined && requiredCaps.canFetch && !caps.canFetch)
            return false;
        if (requiredCaps.canSubmit !== undefined && requiredCaps.canSubmit && !caps.canSubmit)
            return false;
        if (requiredCaps.canValidate !== undefined && requiredCaps.canValidate && !caps.canValidate)
            return false;
        return true;
    }
}
exports.IntegrationRegistry = IntegrationRegistry;
